### ISSUE TITLE
SSR Preview Deployment Resources Not Being Deleted After PR Merge/Close

### ISSUE THREAD

**Issue by EvilTeliportist** [aws-amplify/amplify-hosting#2605]
### Before opening, please confirm:

- [X] I have checked to see if my question is addressed in the [FAQ](https://github.com/aws-amplify/amplify-console/blob/master/FAQ.md).
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-console/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-console/blob/master/CONTRIBUTING.md).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### App Id

d2k26qrgfqnjn5

### Region

us-east-1

### Amplify Console feature

Previews

### Describe the bug

Basically a duplicate of #277 and #2123.

I currently have a React/NextJS/Amplify app running that has built up tons of lambdas, s3 buckets, and cloudfront distros due to previews being opened/merged. There resources are not being deleted after the MR is closed or merged.

For reference we're using GitLab at my team.

### Expected behavior

The resources associated with the MRs should be deleted when the MR is closed/merged.

### Reproduction steps

1. Make a repo with GitLab.
2. Setup a basic NextJS React app.
3. Setup Amplify with previews.
4. Open a MR, make sure preview is deployed.
5. Close or merge the MR.
Resources associated with the MR have not been deleted.

### Build Settings

```yaml
version: 1
frontend:
  phases:
    preBuild:
      commands:
        - npm ci
    build:
      commands:
        - npm run build
  artifacts:
    baseDirectory: .next
    files:
      - '**/*'
  cache:
    paths:
      - node_modules/**/*
```

**Comment by hloriana** [comment]
Hi @EvilTeliportist 👋🏽, thank you for opening this. This is a bug that the team is aware of and are prioritizing to be fixed.

**Comment by EvilTeliportist** [comment]
```
2022-02-25T06:38:43 [INFO]: Beginning deployment for application d2k26qrgfqnjn5, branch:pr-36, buildId 0000000002
2022-02-25T06:38:44 [INFO]: Deploying SSR Resources. Distribution ID: E1E06ZINUIOEU8. This may take a few minutes...
2022-02-25T06:38:44 [INFO]: Deployed the following resources to your account:
2022-02-25T06:38:45 [INFO]: - CloudFront Domain ID: d19uuxso50gao6
2022-02-25T06:38:45 [INFO]: - SSR Lambda@Edge: 9w8i4qo-wyw1thm
2022-02-25T06:38:45 [INFO]: - Image Optimization Lambda@Edge: 9w8i4qo-sup3k6e
2022-02-25T06:38:45 [INFO]: - S3 Bucket: 9w8i4qo-372l4l
2022-02-25T06:38:45 [INFO]: Deployment complete
```

Here's an example deployment log from Amplify. Is there any way for me to get the CloudFront Domain ID, SSR Lambda IDs, and S3 Bucket ID through the AWS Amplify CLI? If so, we could automate this process in our pipeline while you guys are working on a fix.

**Comment by julesrenaud** [comment]
Hope I'm not paying for these deployments...

**Comment by yicru** [comment]
Is there any update on this?
Because of this issue, CloudFront Distribution is not deleted and the build fails.
I have set up a CI to remove distributions with disabled status, but the status remains enabled and does not seem to be working.
```
2022-03-29T14:44:22.509Z [INFO]: ○  (Static)  automatically rendered as static HTML (uses no initial props)
2022-03-29T14:44:22.734Z [INFO]: Done in 59.54s.
2022-03-29T14:44:22.738Z [INFO]: Starting SSR Build...
2022-03-29T14:48:54.890Z [ERROR]: TooManyDistributionsWithLambdaAssociations: Completing this action would cause the maximum number of distributions with Lambda function associations per owner to be exceeded (maximum allowed is 25).
                                  at Request.extractError (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/protocol/rest_xml.js:53:29)
                                  at Request.callListeners (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/sequential_executor.js:106:20)
                                  at Request.emit (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
                                  at Request.emit (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/request.js:686:14)
                                  at Request.transition (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/request.js:22:10)
                                  at AcceptorStateMachine.runTo (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/state_machine.js:14:12)
                                  at /root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/state_machine.js:26:10
                                  at Request.<anonymous> (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/request.js:38:9)
                                  at Request.<anonymous> (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/request.js:688:12)
                                  at Request.callListeners (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/sequential_executor.js:116:18)
                                  at Request.emit (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
                                  at Request.emit (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/request.js:686:14)
                                  at Request.transition (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/request.js:22:10)
                                  at AcceptorStateMachine.runTo (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/state_machine.js:14:12)
                                  at /root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/state_machine.js:26:10
                                  at Request.<anonymous> (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/request.js:38:9)
                                  at Request.<anonymous> (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/request.js:688:12)
                                  at Request.callListeners (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/sequential_executor.js:116:18)
                                  at callNextListener (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/sequential_executor.js:96:12)
                                  at IncomingMessage.onEnd (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/event_listeners.js:335:13)
                                  at IncomingMessage.emit (events.js:412:35)
                                  at IncomingMessage.emit (domain.js:475:12)
                                  at endReadableNT (internal/streams/readable.js:1334:12)
                                  at processTicksAndRejections (internal/process/task_queues.js:82:21) {
                                  code: 'TooManyDistributionsWithLambdaAssociations',
                                  time: 2022-03-29T14:48:54.417Z,
                                  requestId: 'd7950dcd-4ed1-40b4-809f-0907720c9ba5',
                                  statusCode: 400,
                                  retryable: false,
                                  retryDelay: 55.43543223898651
                                  }
2022-03-29T14:48:54.890Z [ERROR]: TooManyDistributionsWithLambdaAssociations
Terminating logging...
```

**Comment by Alek-S** [comment]
We experienced the Cloudfront issue as well due to resources not being cleaned up and on preview PRs. In our case, a team member was unable to push code for a bug fix due to build failures after we hit our account caps. It required manual intervention to clean up some of them to unblock the PR, but it is frankly hard to identify which resources are from which amplify projects and/or PRs without manually tracing through everything.

This also looks to be related to issues #2655 and #2424

**Comment by arbass22** [comment]
> We experienced the Cloudfront issue as well due to resources not being cleaned up and on preview PRs. In our case, a team member was unable to push code for a bug fix due to build failures after we hit our account caps. It required manual intervention to clean up some of them to unblock the PR, but it is frankly hard to identify which resources are from which amplify projects and/or PRs without manually tracing through everything.

Agreed, even if I wanted to create an external process to delete old resources there is no easy way to reliably identify them. I'd love to add a tag to all all of the Amplify managed resources so I know they are Amplify deployment-related.

**Comment by arbass22** [comment]
> Is there any update on this? Because of this issue, CloudFront Distribution is not deleted and the build fails. I have set up a CI to remove distributions with disabled status, but the status remains enabled and does not seem to be working.
> 
> ```
> 2022-03-29T14:44:22.509Z [INFO]: ○  (Static)  automatically rendered as static HTML (uses no initial props)
> 2022-03-29T14:44:22.734Z [INFO]: Done in 59.54s.
> 2022-03-29T14:44:22.738Z [INFO]: Starting SSR Build...
> 2022-03-29T14:48:54.890Z [ERROR]: TooManyDistributionsWithLambdaAssociations: Completing this action would cause the maximum number of distributions with Lambda function associations per owner to be exceeded (maximum allowed is 25).
>                                   at Request.extractError (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/protocol/rest_xml.js:53:29)
>                                   at Request.callListeners (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/sequential_executor.js:106:20)
>                                   at Request.emit (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
>                                   at Request.emit (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/request.js:686:14)
>                                   at Request.transition (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/request.js:22:10)
>                                   at AcceptorStateMachine.runTo (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/state_machine.js:14:12)
>                                   at /root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/state_machine.js:26:10
>                                   at Request.<anonymous> (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/request.js:38:9)
>                                   at Request.<anonymous> (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/request.js:688:12)
>                                   at Request.callListeners (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/sequential_executor.js:116:18)
>                                   at Request.emit (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
>                                   at Request.emit (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/request.js:686:14)
>                                   at Request.transition (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/request.js:22:10)
>                                   at AcceptorStateMachine.runTo (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/state_machine.js:14:12)
>                                   at /root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/state_machine.js:26:10
>                                   at Request.<anonymous> (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/request.js:38:9)
>                                   at Request.<anonymous> (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/request.js:688:12)
>                                   at Request.callListeners (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/sequential_executor.js:116:18)
>                                   at callNextListener (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/sequential_executor.js:96:12)
>                                   at IncomingMessage.onEnd (/root/.//node_modules/@sls-next/aws-cloudfront/node_modules/aws-sdk/lib/event_listeners.js:335:13)
>                                   at IncomingMessage.emit (events.js:412:35)
>                                   at IncomingMessage.emit (domain.js:475:12)
>                                   at endReadableNT (internal/streams/readable.js:1334:12)
>                                   at processTicksAndRejections (internal/process/task_queues.js:82:21) {
>                                   code: 'TooManyDistributionsWithLambdaAssociations',
>                                   time: 2022-03-29T14:48:54.417Z,
>                                   requestId: 'd7950dcd-4ed1-40b4-809f-0907720c9ba5',
>                                   statusCode: 400,
>                                   retryable: false,
>                                   retryDelay: 55.43543223898651
>                                   }
> 2022-03-29T14:48:54.890Z [ERROR]: TooManyDistributionsWithLambdaAssociations
> Terminating logging...
> ```

I considered doing this, but would also need to delete S3 buckets and Lambdas from old deployments as well. And those don't have the concept of "disabled" so it is hard to identify what can be deleted

**Comment by hloriana** [comment]
The team is actively working to fix this bug. We understand manually deleting these resources is not ideal and we are working to fix this issue ASAP. I will close this issue as a duplicate of #2424. Please enable notifications for that issue so you will be notified when the fix has been rolled out.
